panic on memoryStorage.Term

3-node cluster on AWS

```
2014/12/03 19:40:17 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:17 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:17 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:17 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:17 etcdserver: start to snapshot (applied: 6250891, lastsnap: 6237407)
2014/12/03 19:40:17 etcdserver: compacted log at index 6250891
2014/12/03 19:40:17 etcdserver: saved snapshot at index 6250891
2014/12/03 19:40:18 etcdserver: start to snapshot (applied: 6261243, lastsnap: 6250891)
2014/12/03 19:40:18 etcdserver: compacted log at index 6261243
2014/12/03 19:40:18 etcdserver: saved snapshot at index 6261243
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314169] rejected msgApp [logterm: 25, index: 6314169] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314170] rejected msgApp [logterm: 25, index: 6314170] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314171] rejected msgApp [logterm: 25, index: 6314171] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314168] rejected msgApp [logterm: 25, index: 6314168] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314173] rejected msgApp [logterm: 25, index: 6314173] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314172] rejected msgApp [logterm: 25, index: 6314172] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314175] rejected msgApp [logterm: 25, index: 6314175] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314174] rejected msgApp [logterm: 25, index: 6314174] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314176] rejected msgApp [logterm: 25, index: 6314176] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314179] rejected msgApp [logterm: 25, index: 6314179] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314178] rejected msgApp [logterm: 25, index: 6314178] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314177] rejected msgApp [logterm: 25, index: 6314177] from 79bfc23022462c9d
2014/12/03 19:40:29 raft: edc4c934935a1e65 [logterm: 0, index: 6314180] rejected msgApp [logterm: 25, index: 6314180] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314181] rejected msgApp [logterm: 25, index: 6314181] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314182] rejected msgApp [logterm: 25, index: 6314182] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314183] rejected msgApp [logterm: 25, index: 6314183] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314200] rejected msgApp [logterm: 25, index: 6314200] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314201] rejected msgApp [logterm: 25, index: 6314201] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314202] rejected msgApp [logterm: 25, index: 6314202] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314203] rejected msgApp [logterm: 25, index: 6314203] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314372] rejected msgApp [logterm: 25, index: 6314372] from 79bfc23022462c9d
2014/12/03 19:40:30 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:30 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314372] rejected msgApp [logterm: 25, index: 6314372] from 79bfc23022462c9d
2014/12/03 19:40:30 raft: edc4c934935a1e65 [logterm: 0, index: 6314373] rejected msgApp [logterm: 25, index: 6314373] from 79bfc23022462c9d
2014/12/03 19:40:30 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:30 sender: reach the maximal serving to http://172.31.19.177:7001/raft
2014/12/03 19:40:30 raft: edc4c934935a1e65 [commit: 6270819, lastindex: 6278196, lastterm: 25] starts to restore snapshot [index: 6312136, term: 25]
2014/12/03 19:40:30 raftlog: log [committed=6278196, applied=6270819, unstable.offset=6270819, len(unstable.Entries)=1] starts to restore snapshot [index: 6312136, term: 25]
2014/12/03 19:40:30 raft: edc4c934935a1e65 restored progress of 64d55225cc7f3340 [match: 0, next: 6312137]
2014/12/03 19:40:30 raft: edc4c934935a1e65 restored progress of 79bfc23022462c9d [match: 0, next: 6312137]
2014/12/03 19:40:30 raft: edc4c934935a1e65 restored progress of edc4c934935a1e65 [match: 6312136, next: 6312137]
2014/12/03 19:40:30 raft: edc4c934935a1e65 [commit: 6270819] restored snapshot [index: 6312136, term: 25]
panic: runtime error: index out of range

goroutine 20 [running]:
runtime.panic(0x7afc40, 0xa16f9c)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/raft.(*MemoryStorage).Term(0xc20815e000, 0x5fcc34, 0x0, 0x0, 0x0)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/storage.go:116 +0x11e
github.com/coreos/etcd/raft.(*raftLog).term(0xc208169ef0, 0x5fcc34, 0xf8)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/log.go:210 +0xb5
github.com/coreos/etcd/raft.(*raftLog).matchTerm(0xc208169ef0, 0x5fcc34, 0x19, 0x200)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/log.go:240 +0x31
github.com/coreos/etcd/raft.(*raftLog).maybeAppend(0xc208169ef0, 0x5fcc34, 0x19, 0x0, 0xc2085a6000, 0x1, 0x1, 0x5fcc35, 0xb6)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/log.go:77 +0x64
github.com/coreos/etcd/raft.(*raft).handleAppendEntries(0xc2081c8630, 0x3, 0xedc4c934935a1e65, 0x79bfc23022462c9d, 0x19, 0x19, 0x5fcc34, 0xc2085a6000, 0x1, 0x1, ...)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:437 +0x78
github.com/coreos/etcd/raft.stepFollower(0xc2081c8630, 0x3, 0xedc4c934935a1e65, 0x79bfc23022462c9d, 0x19, 0x19, 0x5fcc34, 0xc2085a6000, 0x1, 0x1, ...)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:558 +0x256
github.com/coreos/etcd/raft.(*raft).Step(0xc2081c8630, 0x3, 0xedc4c934935a1e65, 0x79bfc23022462c9d, 0x19, 0x19, 0x5fcc34, 0xc2085a6000, 0x1, 0x1, ...)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:432 +0xe8
github.com/coreos/etcd/raft.(*node).run(0xc208169ea0, 0xc2081c8630)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:256 +0x44a
created by github.com/coreos/etcd/raft.RestartNode
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:164 +0x250
```

--- a/raft/log.go
+++ b/raft/log.go
@@ -199,7 +199,9 @@ func (l *raftLog) stableSnapTo(i uint64) { l.unstable.stableSnapTo(i) }
 func (l *raftLog) lastTerm() uint64 { return l.term(l.lastIndex()) }
 
 func (l *raftLog) term(i uint64) uint64 {
-	if i > l.lastIndex() {
+	// the valid term range is [index of dummy entry, last index]
+	dummyIndex := l.firstIndex() - 1
+	if i < dummyIndex || i > l.lastIndex() {
 		return 0
 	}
 

--- a/raft/log_test.go
+++ b/raft/log_test.go
@@ -634,7 +634,7 @@ func TestTerm(t *testing.T) {
 	num := uint64(100)
 
 	storage := NewMemoryStorage()
-	storage.ApplySnapshot(pb.Snapshot{Metadata: pb.SnapshotMetadata{Index: offset}})
+	storage.ApplySnapshot(pb.Snapshot{Metadata: pb.SnapshotMetadata{Index: offset, Term: 1}})
 	l := newLog(storage)
 	for i = 1; i < num; i++ {
 		l.append(pb.Entry{Index: offset + i, Term: i})
@@ -645,10 +645,40 @@ func TestTerm(t *testing.T) {
 		w     uint64
 	}{
 		{offset - 1, 0},
-		{offset, 0},
+		{offset, 1},
 		{offset + num/2, num / 2},
 		{offset + num - 1, num - 1},
 		{offset + num, 0},
+	}
+
+	for i, tt := range tests {
+		term := l.term(tt.index)
+		if !reflect.DeepEqual(term, tt.w) {
+			t.Errorf("#%d: at = %d, want %d", i, term, tt.w)
+		}
+	}
+}
+
+func TestTermWithUnstableSnapshot(t *testing.T) {
+	storagesnapi := uint64(100)
+	unstablesnapi := storagesnapi + 5
+
+	storage := NewMemoryStorage()
+	storage.ApplySnapshot(pb.Snapshot{Metadata: pb.SnapshotMetadata{Index: storagesnapi, Term: 1}})
+	l := newLog(storage)
+	l.restore(pb.Snapshot{Metadata: pb.SnapshotMetadata{Index: unstablesnapi, Term: 1}})
+
+	tests := []struct {
+		index uint64
+		w     uint64
+	}{
+		// cannot get term from storage
+		{storagesnapi, 0},
+		// cannot get term from the gap between storage ents and unstable snapshot
+		{storagesnapi + 1, 0},
+		{unstablesnapi - 1, 0},
+		// get term from unstable snapshot index
+		{unstablesnapi, 1},
 	}
 
 	for i, tt := range tests {

